_overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,408 @@
+---
+title: FP Modeling
+type: section
+description: This chapter provides an introduction to FP domain modeling with Scala 3.
+num: 22
+previous-page: domain-modeling-oop
+next-page: methods-intro
+---
+
+
+This chapter provides an introduction to functional programming (FP) style domain modeling in Scala 3. When modeling the world around us with FP, you typically use these Scala constructs:
+
+- Enumerations
+- Case classes
+- Traits
+
+
+## Introduction
+
+In FP, the *data* and the *operators on that data* are two separate things; you aren’t forced to encapsulate them together like you do with OOP.
+
+The concept is like numerical algebra. When you think about whole numbers whose values are greater than or equal to zero, you have a *set* of possible values that looks like this:
+
+````
+0, 1, 2 ... Int.MaxInt
+````
+
+Ignoring the division of whole numbers, the possible *operators* on those values are:
+
+````
++, -, *
+````
+
+An FP design is implemented in a similar way:
+
+- You have a set of values
+- You have a collection of operators that work on those values
+
+In this chapter we’ll model the data and operations for a “pizza” in a pizza store. You’ll see how to implement the “data” portion of the Scala/FP model, and then you’ll see several different ways you can organize the operations on that data.
+
+
+
+## Modeling the data
+
+Modeling the “data” portion of a domain model in Scala is simple:
+
+- Create ADTs and GADTs with the `enum` construct
```

Comment:
A quick introduction to ADT or GADT might be useful before this

Reply:
I don’t know if it’s enough, but I just added a link to the ADT page at the top of this page.

Reply:
I’m resolving this just to make it easier for me to find unresolved issues here, but please re-open this if you think it needs to be addressed more.